docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,182 @@
+on hypervisor, sandbox represents a virtual machine. For others, it can be a
+container holding the namespaces.
+
+In short, a PodSandbox should have the following features.
+
+ * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
+   support additional security features.
+ * **Resource requirements**: A sandbox should implement pod-level resource
+   requirements.
+
+A container in a PodSandbox maps to a application in the Pod Spec. For Linux
+containers, they are expected to share at least network and IPC namespaces,
+with sharing more namespaces discussed in [#1615](https://issues.k8s.io/1615).
+
+
+Below is an example of the proposed interfaces.
+```go
+// PodSandboxManager contains basic operations for sandbox.
+type PodSandboxManager interface {
+	Create(config *PodSandboxConfig) (string, error)
+	Delete(id string) (string, error)
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	Inspect(id string) PodSandbox
+}
+
+// ContainerRuntime contains basic operations for containers.
+type ContainerRuntime interface {
+    Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxIDng) (string, error)
+    Start(id string) error
+    Stop(id string, timeout int) error
+    Remove(id string) error
+    List(filter ContainerFilter) ([]ContainerListItem, error)
+    Inspect(id string) (Container, error)
+}
+
+// ContainerCommandExecutor provides methods to run commands in a container.
+type ContainerCommandExecutor interface {
+	Exec(id string, cmd []string, streamOpts StreamOptions) error
+}
+
+// ImageOperations contains image-related operations.
+type ImageOperations interface {
+	List() ([]Image, error)
+	Pull(image ImageSpec, auth AuthConfig) error
+	Remove(image ImageSpec) error
+	Inspect(image ImageSpec) (Image, error)
+}
+
+type PodSandboxMetricsGetter interface {
+    ContainerMetrics(id string) (ContainerMetrics, error)
+}
+
+type ContainerMetricsGetter interface {
+    ContainerMetrics(id string) (ContainerMetrics, error)
+}
+
+type ImageMetricsGetter interface {
+    ImageMetrics(id string) (ImageMetrics, error)
+}
+```
+
+### Pod Lifecycle
+
+The sandbox’s lifecycle is decoupled from the containers, i.e., a sandbox
+is created before any containers, and can exist after all containers in it have
+terminated.
+
+Assume there is a pod with a single container C. To start a pod:
+```
+  create sandbox Foo --> create container C --> start container C
+```
+
+To delete a pod:
+```
+  stop container C --> remove container C --> delete sandbox Foo
+```
+
+Kubelet is responsible for creating, starting the containers based on the
+pod-level restart policy. E.g., if container C dies unexpectedly, Kubelet
+will simply create and start the container again in sandbox Foo.
+
+Kubelet is also responsible for gracefully terminating all the containers
+in the sandbox before deleting the sandbox. If Kubelet chooses to delete
+the sandbox with running containers in it, those containers may be forcibly
```

Comment:
Replace `may` with `should`. We want the implementer of this interface to kill the containers no matter what to satisfy a `PodSandbox` Delete request.